src/Fields/fieldvector.jl

```diff
@@ -203,6 +203,93 @@ end
     parent(getfield(_values(fv), symb))
 @inline transform_broadcasted(x, symb, axes) = x
+
+@inline function first_fieldvector_in_bc(args::Tuple, rargs...)
+    x1 = first_fieldvector_in_bc(args[1], rargs...)
+    x1 isa FieldVector && return x1
+    return first_fieldvector_in_bc(Base.tail(args), rargs...)
+end
+
+@inline first_fieldvector_in_bc(args::Tuple{Any}, rargs...) =
+    first_fieldvector_in_bc(args[1], rargs...)
+@inline first_fieldvector_in_bc(args::Tuple{}, rargs...) = nothing
+@inline first_fieldvector_in_bc(x) = nothing
+@inline first_fieldvector_in_bc(x::FieldVector) = x
+
+@inline first_fieldvector_in_bc(
+    bc::Base.Broadcast.Broadcasted{FieldVectorStyle},
+) = first_fieldvector_in_bc(bc.args)
+
+@inline _is_diagonal_bc_args(
+    truesofar,
+    ::Type{TStart},
+    args::Tuple,
+    rargs...,
+) where {TStart} =
+    truesofar &&
+    _is_diagonal_bc(truesofar, TStart, args[1], rargs...) &&
+    _is_diagonal_bc_args(truesofar, TStart, Base.tail(args), rargs...)
+
+@inline _is_diagonal_bc_args(
+    truesofar,
+    ::Type{TStart},
+    args::Tuple{Any},
+    rargs...,
+) where {TStart} =
+    truesofar && _is_diagonal_bc(truesofar, TStart, args[1], rargs...)
+@inline _is_diagonal_bc_args(
+    truesofar,
+    ::Type{TStart},
+    args::Tuple{},
+    rargs...,
+) where {TStart} = truesofar
+
+@inline function _is_diagonal_bc(
+    truesofar,
+    ::Type{TStart},
+    bc::Base.Broadcast.Broadcasted{FieldVectorStyle},
+) where {TStart}
+    return truesofar && _is_diagonal_bc_args(truesofar, TStart, bc.args)
+end
+
+@inline _is_diagonal_bc(
+    truesofar,
+    ::Type{TStart},
+    ::TStart,
+) where {TStart <: FieldVector} = true
+@inline _is_diagonal_bc(
+    truesofar,
+    ::Type{TStart},
+    x::FieldVector,
+) where {TStart} = false
+@inline _is_diagonal_bc(truesofar, ::Type{TStart}, x) where {TStart} = truesofar
+
+# Find the first fieldvector in the broadcast expression (BCE),
+# and compare against every other fieldvector in the BCE
+@inline is_diagonal_bc(bc::Base.Broadcast.Broadcasted{FieldVectorStyle}) =
+    _is_diagonal_bc_args(true, typeof(first_fieldvector_in_bc(bc)), bc.args)
+
+# Specialize on FieldVectorStyle to avoid inference failure
+# in fieldvector broadcast expressions:
+# https://github.com/JuliaArrays/BlockArrays.jl/issues/310
+function Base.Broadcast.instantiate(
+    bc::Base.Broadcast.Broadcasted{FieldVectorStyle},
+)
+    if bc.axes isa Nothing # Not done via dispatch to make it easier to extend instantiate(::Broadcasted{Style})
+        axes = Base.Broadcast.combine_axes(bc.args...)
+    else
+        axes = bc.axes
+        # Base.Broadcast.check_broadcast_axes is type-unstable
+        # for broadcast expressions with multiple fieldvectors.
+        # So, let's statically elide this when we have "diagonal"
+        # broadcast expressions:
+        if !is_diagonal_bc(bc)
+            Base.Broadcast.check_broadcast_axes(axes, bc.args...)
```

Comment:
Can we just throw an error when this isn't true? The example for this in the unit tests can't be materialized, because materializing it would result in a `Field` broadcast error for mismatched spaces. Why not throw an error for this at the `FieldVector` broadcast level?

Reply:
It's a good question, I suppose that we could (I have no objection). The only reason I didn't was because, in theory, we could support broadcasting over different field vectors (we would obviously need to define rules). That said, it's likely not going to ever happen. Thoughts @Sbozzolo / @sriharshakandala ?

Reply:
I guess we might want to allow broadcasting over `FieldVector`s with the same spaces and data structures, but with different basetypes? Like adding a `FieldVector` of `Float32`s to a `FieldVector` of `Dual`s. But that would require dealing with promotion rules, and I haven't seen any need for it yet. We could also just implement it as an extension of `is_diagonal_bc`, with more complicated type checking.

Reply:
True. We can always add the error and remove it if we find a need.

Reply:
I've added an error for when it's not diagonal